File: 151.反转字符串中的单词.c
/*
 * @lc app=leetcode.cn id=151 lang=c
 *
 * [151] 反转字符串中的单词
 */

// @lc code=start

void reverse(char *s, int start, int end)
{
  if (start >= end)
    return;
  int i, j;
  char temp;
  i = start;
  j = end;
  while (i < j)
  {
    temp = s[i];
    s[i] = s[j];
    s[j] = temp;
    i++;
    j--;
  }
}

char *reverseWords(char *s)
{
  // int i, j, len;
  // len = strlen(s);

  // i = 0;
  // j = 0;
  // while (s[j] == ' ')
  //   j++;

  // while (s[j] != '\0')
  // {
  //   if (j > 1 && s[j] == ' ' && s[j - 1] == ' ')
  //     j++;
  //   else
  //   {
  //     s[i] = s[j++];
  //     i++;
  //   }
  // }
  // if (i > 1 && s[i - 1] == ' ')
  //   i--;
  // s[i] = '\0';

  // len = i;
  // reverse(s, 0, len - 1);
  // i = 0;
  // j = i;
  // while (s[i] != '\0')
  // {
  //   if (s[i] == ' ')
  //   {
  //     reverse(s, j, i - 1);
  //     j = i + 1;
  //   }
  //   i++;
  // }
  // reverse(s, j, i - 1);

  // return s;

  int i, j, len;
  len = strlen(s);
  // 去除首空格
  for (i = 0; s[i] == ' '; i++, len--)
    ;
  // 去除中间空格
  for (j = 0; i < strlen(s); i++)
  {
    /* code */
    if (i > 0 && s[i] == ' ' && s[i - 1] == ' ')
    {
      len--;
      continue;
    }

    s[j++] = s[i];
  }

  // 去除末尾空格
  if (s[j - 1] == ' ')
  {
    len--;
    s[j - 1] = '\0';
  }
  else
    s[j] = '\0';

  j = 0;
  reverse(s, j, len - 1);
  for (i = 0; s[i] != '\0'; i++)
  {
    if (s[i] == ' ')
    {
      reverse(s, j, i - 1);
      j = i + 1;
    }
  }
  reverse(s, j, i - 1);

  return s;
}
// @lc code=end

File: store/151.反转字符串中的单词.js
/*
 * @lc app=leetcode.cn id=151 lang=javascript
 *
 * [151] 反转字符串中的单词
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
const reverse = (start, end, arr) => {
  let temp;
  while (start < end) {
    temp = arr[start];
    arr[start] = arr[end];
    arr[end] = temp;
    start++;
    end--;
  }
};
var reverseWords = function (s) {
  // 法一：水方法
  // let item,
  //   sArr = [],
  //   i;
  // item = "";
  // for (i = 0; i < s.length; i++) {
  //   if (i > 0 && s[i - 1] === " " && s[i] === " ") continue;
  //   if (s[i] !== " ") item += s[i];
  //   if (s[i] !== " " && (s[i + 1] === " " || i + 1 === s.length)) {
  //     sArr.push(item);
  //     item = "";
  //   }
  // }
  // reverse(0, sArr.length - 1, sArr);
  // return sArr.join(" ");
  // 法二：空间复杂度为O(1)
  let slow, fast, lastIndex;
  slow = fast = lastIndex = 0;
  s = Array.from(s);
  while (s[fast] === " ") fast++;
  for (; fast < s.length; fast++) {
    if (s[fast] === " " && (s[fast + 1] === " " || fast + 1 === s.length))
      continue;
    s[slow++] = s[fast];
  }

  s.length = slow;
  reverse(0, s.length - 1, s);
  s.forEach((item, index) => {
    if (item === " ") {
      reverse(lastIndex, index - 1, s);
      lastIndex = index + 1;
    }
    if (index === s.length - 1) reverse(lastIndex, index, s);
  });
  return s.join("");
};
// @lc code=end
